### PR TITLE
Validate the start and stop date of a page

### LINKED ISSUE #812: Validate stop date > start date
**Affected version(s)**

I think all versions will be affected.

**Description**

Actually the function _show from/show until_ is working just for the following cases:

1. _show from_ is activated
2. _show until_ is activated
3. _both options_ are activated, but _show until_ must be a higher date than _show from_

The following option is not working:
4. _both options_ are activated, but _show until_ is a smaller date than _show from_ 

**How to reproduce**

I have a text element (I think, the problem also exists for articles and pages), which is online, but I want to hide it for a day:
_show from_ is **2019-10-04 00:01**
_show until_ is **2019-10-03 00:01**

I expect, that the text is shown until 2019-10-03 00:01, hidden the whole day of 2019-10-03 and visible again until 2019-10-04 00:01.

But the result is: the text is never visible.

### PR DESCRIPTION
**Pull request by leofeyer**
Fixes #812 

If we agree that the changes are correct, we have to apply them to articles, news, events etc. as well.

**TODO**

- [x] Unit tests
- [x] Add the save callback to other start/stop fields as well

### CODE REVIEW

**Comment by aschempp** [comment]
Can we make this a generic callback listener that can be registered on any DCA field? Every DCA will otherwise have to copy the whole code.

**Comment by leofeyer** [comment]
IDK. Where would you put that generic listener? And how would you configure it (e.g. field names)?

**Review by aschempp — CHANGES_REQUESTED** [review]
I have added an actual event listener to implement the same code as yours does. The event listener can handle any table, and there can be additional instances if the fields are not called `start` and `stop`.

When merging this into Contao 4.8, we can also use the translator instead of globals and use services tags instead of adjusting the DCA to register the callbacks.

Your code is still there, one should be cleaned up before merging 😉

**Comment by leofeyer** [comment]
I see and I agree. However, I would prefer to add this as a new feature in Contao 4.9 then so we do not have to do the work twice (the listener also needs a unit test).

**Comment by aschempp** [comment]
You mean no validation in Contao 4.4? I would be fine with that.

**Comment by Mynyx** [comment]
The title of the ticket #812 is "Validate stop date > start date", but the implementation is stop date >= start date.

**Comment by fritzmg** [comment]
@Mynyx what's your point? ;) Why should start == stop be valid?

**Comment by Mynyx** [comment]
> @Mynyx what's your point? ;) Why should start == stop be valid?

It should not be valid, but the implementations in this pull request would accept it as valid.

**Comment by leofeyer** [comment]
@aschempp I have adjusted the PR according to your suggestions. Can you please check again before I continue working on it?

**Comment by aschempp** [comment]
Why did you not use the service tags/annotations to register the callbacks?

**Comment by leofeyer** [comment]
Do you mean the `@Callback` annotation?

**Comment by m-vo** [comment]
... or the service tagged with `name: 'contao.callback', callback: '...'`

**Comment by aschempp** [comment]
Yes, either of these would be better to lead as an example 😊

**Comment by leofeyer** [comment]
Changed in a55254d5e8c407abf74eda628cb99aed6d60a623. Any other remarks?

File: core-bundle/src/DataContainer/StartStopValidator.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        if (null === $request) {
+            throw new \RuntimeException('The request stack did not contain a request');
+        }
+
+        $stop = $request->request->has($this->stopField)
+            ? strtotime($request->request->get($this->stopField))
+            : $dc->activeRecord->{$this->stopField}
+        ;
+
+        if ($stop && $value > $stop) {
```
Comment (Mynyx) [inline core-bundle/src/DataContainer/StartStopValidator.php]:
```suggestion
        if ($stop && $value >= $stop) {
```

File: core-bundle/src/Resources/contao/languages/en/default.xlf [hunk 2]
```diff
@@ -230,6 +230,12 @@
       <trans-unit id="ERR.moduleWithoutColumn">
         <source>Please assign all modules to a column!</source>
       </trans-unit>
+      <trans-unit id="ERR.startAfterStop">
+        <source>The start time cannot be after the end time!</source>
```
Comment (Mynyx) [inline core-bundle/src/Resources/contao/languages/en/default.xlf]:
```suggestion
        <source>The start time must be before the end time!</source>
```

File: core-bundle/src/Resources/contao/languages/en/default.xlf [hunk 3]
```diff
@@ -230,6 +230,12 @@
...
+      </trans-unit>
+      <trans-unit id="ERR.stopBeforeStart">
+        <source>The end time cannot be before the start time!</source>
```
Comment (Mynyx) [inline core-bundle/src/Resources/contao/languages/en/default.xlf]:
```suggestion
        <source>The end time must be after the start time!</source>
```

File: core-bundle/src/DataContainer/StartStopValidator.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        if (null === $request) {
+            throw new \RuntimeException('The request stack did not contain a request');
+        }
+
+        $start = $request->request->has($this->startField)
+            ? strtotime($request->request->get($this->startField))
+            : $dc->activeRecord->{$this->startField}
+        ;
+
+        if ($start && $start > $value) {
```
Comment (Mynyx) [inline core-bundle/src/DataContainer/StartStopValidator.php]:
```suggestion
        if ($start && $start >= $value) {
```

**Review by Mynyx — CHANGES_REQUESTED** [review]
start == stop should not be valid.

**Comment by leofeyer** [comment]
> start == stop should not be valid.

Changed in 51dc5db4c9df62bb965b854ff71d5917f4cc9962.

**Comment by hhogrebe** [comment]
Maybe the translation ids should also be amended.

**Comment by Mynyx** [comment]
Maybe the translation IDs should also be amended.

**Comment by leofeyer** [comment]
@contao/developers This PR is RTM now.

**Review by aschempp — COMMENTED** [review]
Can you explain what the `StartStopValidatorPass` is for?

**Comment by leofeyer** [comment]
It adds the calendar and news bundle fields so they can use the callback, too?

**Comment by aschempp** [comment]
So the core bundle is adding listeners for the news and calendar bundle??

**Comment by leofeyer** [comment]
No? Can you please review carefully? There is one `StartStopValidatorPass` in the calendar bundle and one in the news bundle. Zero in the core bundle.

**Comment by aschempp** [comment]
As discussed, we should try to optimize this, by having a DCA flag instead of a compiler pass per bundle. Something like `start => eval => lowerThan => 'stop'`. Or maybe there should be multiple validators?

```
'fields' => [
    'start' => [
        'eval' => ['validate' => 'lt:stop AND gte:foobar']
    ]
```

**Review by aschempp — COMMENTED** [review]
As discussed, we should try to optimize this, by having a DCA flag instead of a compiler pass per bundle. Something like `start => eval => lowerThan => 'stop'`. Or maybe there should be multiple validators?

```
'fields' => [
    'start' => [
        'eval' => ['validate' => 'lt:stop AND gte:foobar']
    ]
```

**Comment by Toflar** [comment]
The Symfony Validator solves this by using the Expression Language ;)

**Comment by Toflar** [comment]
Draft PR for an alternative approach: https://github.com/contao/contao/pull/1134

**Pull request closed without merging**